src/router.js

```diff
@@ -42,12 +43,32 @@ export default function router(history, routes) {
         [lastSaga] = effects;
       }
 
+      if ('beforeRouteChange' in options) {
+        return {
+          value: take(historyChannel),
+          next: BEFORE_HANDLE_LOCATION,
+        };
+      }
+
       return {
         value: take(historyChannel),
         next: HANDLE_LOCATION,
       };
     },
 
+    [BEFORE_HANDLE_LOCATION](location, fsm) {
+      const path = location.pathname;
+      const match = routeMatcher.match(path);
+      if (!match) {
+        return fsm[LISTEN]();
+      }
+
+      return {
+        value: spawn(options.beforeRouteChange, match.params),
+        next: HANDLE_LOCATION,
+      };
+    },
+
     [HANDLE_LOCATION](location, fsm) {
```

Comment:
When we hit this line after BEFORE_HANDLE_LOCATION, we will have lost reference to the `location`. In fact, the `location` argument will be equal to a task object that redux-saga will return for the `spawn` effect. We'll need to probably save a reference to the location. Maybe create a `let currentLocation` with the other `let` declarations above? Then HANDLE_LOCATION will have to use the `currentLocation` if the `location` argument is missing.

We'll also need another state after BEFORE_HANDLE_LOCATION. While the `spawn` is outside the array now, it is still a non-blocking effect in redux-saga, meaning it has the potential to run concurrently with the matched saga. If you add an additional state, you can use the [`join`](https://redux-saga.github.io/redux-saga/docs/api/index.html#jointask) effect on the task to basically wait for the `beforeRouteChange` saga to finish before handling the location with the route saga.